In [50]:
#讀入資料
import pandas as pd
Orignal = pd.read_csv('HW3_hr-analytics.csv')

#判斷格式是否為數字
Format = [Orignal[Orignal.columns[i]].dtype for i in range(Orignal.shape[1])]
FormatIsntNumber = list()
print("Colums's format is not number:", end = ' ')
for i in range(len(Format)):
    if Format[i] == 'O':
        print(Orignal.columns[i], end = ' ')
        FormatIsntNumber.append(Orignal.columns[i])

#判斷是否有缺失
missColums = list(set(Orignal.columns).symmetric_difference(set(Orignal.dropna(axis = 1).columns)))
if len(missColums):
    print("\n\nColums with miss value:", end = ' ')
    for miss in missColums:
        print(miss, end = ' ')
else:
    print("\n\nNo Miss value.")

Colums's format is not number: sales salary 

No Miss value.


In [51]:
#將非數字資料編碼成數字
from sklearn import preprocessing
import numpy as np
Data = Orignal.copy()

for i in range(len(FormatIsntNumber)):
    encode = preprocessing.LabelEncoder().fit(Orignal[FormatIsntNumber[i]].dropna())
    if not FormatIsntNumber[i] in missColums:
        Data[FormatIsntNumber[i]] = encode.transform(Orignal[FormatIsntNumber[i]])
    else:
        buffer = encode.transform(Orignal[FormatIsntNumber[i]].dropna())
        index = 0
        result = list()
        for j in range(Orignal.shape[0]):
            if str(Orignal[FormatIsntNumber[i]][j])!='nan':
                result.append(buffer[index])
                index += 1
            else:
                result.append(np.nan)
        Data[FormatIsntNumber[i]] = result

print("Encode Data")
Data

Encode Data


satisfaction_level  last_evaluation  number_project  \
0                    0.38             0.53               2   
1                    0.80             0.86               5   
2                    0.11             0.88               7   
3                    0.72             0.87               5   
4                    0.37             0.52               2   
...                   ...              ...             ...   
14994                0.40             0.57               2   
14995                0.37             0.48               2   
14996                0.37             0.53               2   
14997                0.11             0.96               6   
14998                0.37             0.52               2   

       average_montly_hours  time_spend_company  Work_accident  left  \
0                       157                   3              0     1   
1                       262                   6              0     1   
2                       272                   4              0     1   
3                       223                   5              0     1   
4                       159                   3              0     1   
...                     ...                 ...            ...   ...   
14994                   151                   3              0     1   
14995                   160                   3              0     1   
14996                   143                   3              0     1   
14997                   280                   4              0     1   
14998                   158                   3              0     1   

       promotion_last_5years  sales  salary  
0                          0      7       1  
1                          0      7       2  
2                          0      7       2  
3                          0      7       1  
4                          0      7       1  
...                      ...    ...     ...  
14994                      0      8       1  
14995                      0      8       1  
14996                      0      8       1  
14997                      0      8       1  
14998                      0      8       1  

[14999 rows x 10 columns]

In [52]:
if len(missColums):
    Data = Data.fillna(Data.mean())
    print("fill miss Data")
else:
    print("No Miss Value.")

No Miss Value.


In [53]:
from sklearn.model_selection import train_test_split
trainData, testData, a, b = train_test_split(Data, range(Data.shape[0]), test_size=0.3, random_state=None)
print("Train Data")
trainData

Train Data


satisfaction_level  last_evaluation  number_project  \
11646                0.88             0.51               3   
2178                 0.21             0.82               4   
5650                 0.99             0.82               6   
13622                0.94             0.51               6   
11020                0.65             0.91               4   
...                   ...              ...             ...   
2320                 0.92             0.68               5   
7452                 0.74             0.56               3   
4854                 0.79             0.69               4   
12447                0.10             0.84               5   
5996                 1.00             0.94               3   

       average_montly_hours  time_spend_company  Work_accident  left  \
11646                   211                   7              0     0   
2178                    207                   5              0     0   
5650                    185                   4              1     0   
13622                   239                   5              0     0   
11020                   224                   2              1     0   
...                     ...                 ...            ...   ...   
2320                    236                   2              0     0   
7452                    125                   6              0     0   
4854                    209                   2              0     0   
12447                   298                   4              0     1   
5996                    272                   3              0     0   

       promotion_last_5years  sales  salary  
11646                      0      7       2  
2178                       0      9       0  
5650                       0      7       0  
13622                      0      7       2  
11020                      0      7       1  
...                      ...    ...     ...  
2320                       0      3       1  
7452                       0      8       2  
4854                       0      8       2  
12447                      0      0       1  
5996                       0      9       1  

[10499 rows x 10 columns]

In [54]:
print("Test Data")
testData

Test Data


satisfaction_level  last_evaluation  number_project  \
14978                0.40             0.47               2   
6331                 0.75             0.49               2   
14746                0.37             0.56               2   
14867                0.79             0.86               5   
1785                 0.09             0.96               7   
...                   ...              ...             ...   
2017                 0.41             0.84               6   
12732                0.85             0.58               3   
1766                 0.44             0.56               2   
13374                0.77             0.75               4   
12192                0.74             0.91               4   

       average_montly_hours  time_spend_company  Work_accident  left  \
14978                   144                   3              0     1   
6331                    173                   3              1     0   
14746                   156                   3              0     1   
14867                   235                   5              0     1   
1785                    264                   4              0     1   
...                     ...                 ...            ...   ...   
2017                    191                   6              0     0   
12732                   226                   2              0     1   
1766                    141                   3              0     1   
13374                   191                   3              0     0   
12192                   232                   5              0     1   

       promotion_last_5years  sales  salary  
14978                      0      3       2  
6331                       0      1       1  
14746                      0      7       2  
14867                      0      8       2  
1785                       0      2       2  
...                      ...    ...     ...  
2017                       0      3       2  
12732                      0      8       2  
1766                       0      7       1  
13374                      0      0       2  
12192                      0      9       2  

[4500 rows x 10 columns]

In [55]:
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.neural_network import MLPClassifier
from sklearn.metrics import confusion_matrix
import numpy as np
import matplotlib.pyplot as plt


def fromDataframeToPrediectResult(comment, trainData, testData):
    trainLabel, testLabel = list(trainData['left']), list(testData['left'])
    trainData = np.array(trainData.drop(['left'], axis = 1))
    testData = np.array(testData.drop(['left'], axis = 1))
    #訓練
    Model = DecisionTreeClassifier(random_state=0).fit(trainData, trainLabel) if comment == "Decision" else MLPClassifier(random_state=1, max_iter=300).fit(trainData, trainLabel)
    #預測
    ModelResult = Model.predict(testData)
    #統計結果
    
    tn, fp, fn, tp = confusion_matrix(testLabel, ModelResult).ravel()
    accuracy = (tp+tn)/(tn+fp+fn+tp)
    print(comment, "Data with accuracy:",round(accuracy*100,2),"%")


fromDataframeToPrediectResult("Neural Network", trainData, testData)

fromDataframeToPrediectResult("Decision", trainData, testData)

Neural Network Data with accuracy: 92.16 %
Decision Data with accuracy: 97.96 %


# 和前一次之Logistic Regression進行比較

上次的作業Logistic Regression的準確率約為75%左右，

這次的結果Neural Network的結果為92%、Decision Tree的結果則為98%。

我認為是因為Logistic Regression所使用的方式是利用一條線的回歸結果來進行分類，對於複雜的資料分布狀況無法將資料區分乾淨，所以準確率較低。

神經網路依照上課內容，則可以進行較為複雜的分割，因此結果上比Logistic Regression；

至於Decision Tree的結果最好，我認為是因為是因為Desicion Tree是透過一層層的二分法來進行分類，這次作業的資料型態也有多個欄位是離散的資料，因此Decision Tree的結果最好。

**註** 上次作業在計算結果時最後計算準確率時，我不慎使用了錯誤的分母(全部的訓練資料)，而實際應該是用測試資料計算正確率

在第六段程式碼中的fromDataframeToPrediectResult最後一行 Data.shape[0] 應該改成 testData.shape[0]